pkg/workflows/internal/v2/sdkimpl/runtime.go

```diff
@@ -0,0 +1,110 @@
+package sdkimpl
+
+import (
+	"errors"
+	"io"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	valuespb "github.com/smartcontractkit/chainlink-common/pkg/values/pb"
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows/sdk/v2"
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows/sdk/v2/pb"
+)
+
+type CallCapabilityFn func(request *pb.CapabilityRequest) ([]byte, error)
+type AwaitCapabilitiesFn func(request *pb.AwaitCapabilitiesRequest, maxResponseSize uint64) (*pb.AwaitCapabilitiesResponse, error)
+
+type RuntimeBase struct {
+	ExecId          string
+	ConfigBytes     []byte
+	MaxResponseSize uint64
+	Call            CallCapabilityFn
+	Await           AwaitCapabilitiesFn
+	Writer          io.Writer
+
+	modeErr error
+}
+
+func (r *RuntimeBase) CallCapability(request *pb.CapabilityRequest) sdk.Promise[*pb.CapabilityResponse] {
+	if r.modeErr != nil {
+		return sdk.PromiseFromResult[*pb.CapabilityResponse](nil, r.modeErr)
+	}
+
+	request.ExecutionId = r.ExecId
+	id, err := r.Call(request)
+	if err != nil {
+		return sdk.PromiseFromResult[*pb.CapabilityResponse](nil, err)
+	}
+
+	return sdk.NewBasicPromise(func() (*pb.CapabilityResponse, error) {
+		awaitRequest := &pb.AwaitCapabilitiesRequest{
+			ExecId: r.ExecId,
+			Ids:    []string{string(id)},
+		}
+		awaitResponse, err := r.Await(awaitRequest, r.MaxResponseSize)
+		if err != nil {
+			return nil, err
+		}
+
+		capResponse, ok := awaitResponse.Responses[string(id)]
```

Comment:
Do we always expect len(Responses) == 1 here? Can you add a comment?

Reply:
In this case we do, but it's possible that we refactor once we add await helpers that let you await multiple values. r.Await in those cases could return a cached response. The only expectation is that the id you request is in the response.

Reply:
nit: It would be cool to have a comment inline with what you just said :)